plugins/goodix-tp/fu-goodixtp-brlb-firmware.c

```diff
@@ -0,0 +1,146 @@
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-goodixtp-brlb-firmware.h"
+#include "fu-goodixtp-common.h"
+
+struct _FuGoodixtpBrlbFirmware {
+	FuGoodixtpFirmware parent_instance;
+};
+
+G_DEFINE_TYPE(FuGoodixtpBrlbFirmware, fu_goodixtp_brlb_firmware, FU_TYPE_GOODIXTP_FIRMWARE)
+
+#define FW_HEADER_SIZE	      512
+#define FW_SUBSYS_INFO_OFFSET 42
+#define FW_SUBSYS_INFO_SIZE   10
+
+gboolean
+fu_goodixtp_brlb_firmware_parse(FuGoodixtpFirmware *self,
+				GBytes *fw,
+				guint8 sensor_id,
+				GError **error)
+{
+	guint32 version;
+	guint16 vid;
+	guint32 fw_offset = FW_HEADER_SIZE;
+	guint32 info_offset = FW_SUBSYS_INFO_OFFSET;
+	guint32 read_cksum;
+	guint32 checksum = 0;
+	guint8 *tmp_data;
+	guint32 firmware_size;
+	guint8 subsys_num;
+	guint8 cfg_ver = 0;
+	gsize bufsz = 0;
+	const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+
+	if (!fu_memread_uint32_safe(buf, bufsz, 0, &firmware_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	firmware_size += 8;
+	if ((gsize)firmware_size < bufsz) {
+		gint cfg_len = bufsz - (gint)firmware_size - 64;
+		guint8 *cfg_data = (guint8 *)buf + firmware_size + 64;
+		fu_goodixtp_add_chunk_data(self, 4, 0x40000, cfg_data, cfg_len);
+		cfg_ver = cfg_data[34];
+		g_debug("config size:%d, config ver:0x%02x", cfg_len, cfg_ver);
+	}
+
+	for (guint i = 8; i < firmware_size; i += 2) {
+		guint16 tmp_val;
+		if (!fu_memread_uint16_safe(buf, bufsz, i, &tmp_val, G_LITTLE_ENDIAN, error))
+			return FALSE;
+		checksum += tmp_val;
+	}
+
+	if (!fu_memread_uint32_safe(buf, bufsz, 4, &read_cksum, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (checksum != read_cksum) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "firmware bin checksum error");
+		return FALSE;
+	}
+
+	if (!fu_memread_uint8_safe(buf, bufsz, 29, &subsys_num, error))
+		return FALSE;
+	if (subsys_num == 0) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE, "subsys_num is 0, exit");
+		return FALSE;
+	}
+
+	for (guint i = 0; i < subsys_num; i++) {
+		guint8 tmp_type;
+		guint32 tmp_size;
+		guint32 tmp_addr;
+		guint32 offset = 0;
+
+		if (!fu_memread_uint8_safe(buf, bufsz, info_offset, &tmp_type, error))
+			return FALSE;
+		if (!fu_memread_uint32_safe(buf,
+					    bufsz,
+					    info_offset + 1,
+					    &tmp_size,
+					    G_LITTLE_ENDIAN,
+					    error))
+			return FALSE;
+		if (!fu_memread_uint32_safe(buf,
+					    bufsz,
+					    info_offset + 5,
+					    &tmp_addr,
+					    G_LITTLE_ENDIAN,
+					    error))
+			return FALSE;
+
+		tmp_data = (guint8 *)buf + fw_offset;
+		fw_offset += tmp_size;
+		info_offset += 10;
+		if (tmp_type == 0x0B || tmp_type == 0x01) {
+			g_debug("skip type[%x] subsystem", tmp_type);
+			continue;
+		}
+
+		for (gint j = 0; j < (gint)(tmp_size / RAM_BUFFER_SIZE); j++) {
+			fu_goodixtp_add_chunk_data(self,
+						   tmp_type,
+						   tmp_addr + offset,
+						   tmp_data + offset,
```

Comment:
this looks a bit risky; i think you need to check fw_offset+offset to see if it's in-range.